pkg/kubelet/mirror_manager.go

```diff
@@ -0,0 +1,165 @@
+	DeleteMirrorPod(string) error
+}
+
+type basicMirrorManager struct {
+	// mirror pods are stored in the kubelet directly because they need to be
+	// in sync with the internal pods.
+	apiserverClient client.Interface
+}
+
+func newBasicMirrorManager(apiserverClient client.Interface) *basicMirrorManager {
+	return &basicMirrorManager{apiserverClient: apiserverClient}
+}
+
+// Creates a mirror pod.
+func (self *basicMirrorManager) CreateMirrorPod(pod *api.BoundPod, hostname string) error {
+	if self.apiserverClient == nil {
+		return nil
+	}
+	podFullName := GetPodFullName(pod)
+	// Create a Pod from BoundPod; this will become redundant once we
+	// deprecate BoundPod completely.
+	// Note that some fields (e.g. creationTimestamp, UID, etc) will be
+	// overwritten by the APIserver.
+	podToCreate := api.Pod{}
+	if err := api.Scheme.Convert(pod, &podToCreate); err != nil {
+		glog.Errorf("Unable to interpret static bound pod as a pod: %v: %+v", err, pod)
+	}
+	// Indicate that the pod should be scheduled to the current node.
+	podToCreate.Spec.Host = hostname
+	podToCreate.Annotations[ConfigMirrorAnnotationKey] = MirrorType
+
+	glog.V(4).Infof("Creating a mirror pod %q", podFullName)
+	_, err := self.apiserverClient.Pods(NamespaceDefault).Create(&podToCreate)
+	if err != nil {
+		glog.Errorf("Failed creating a mirror pod %q: %#v", podFullName, err)
+		return err
+	}
+	return nil
+}
+
+// Deletes a mirror pod.
+func (self *basicMirrorManager) DeleteMirrorPod(podFullName string) error {
+	if self.apiserverClient == nil {
+		return nil
+	}
+	name, namespace, err := parsePodFullName(podFullName)
+	if err != nil {
+		return err
+	}
+	glog.V(4).Infof("Deleting a mirror pod %q", podFullName)
+	if err := self.apiserverClient.Pods(namespace).Delete(name); err != nil {
+		glog.Errorf("Failed deleting a mirror pod %q: %v", podFullName, err)
+	}
+	return nil
+}
+
+// Delete all orphaned mirror pods.
+func (kl *Kubelet) deleteOrphanedMirrorPods(pods []api.BoundPod, mirrorPods util.StringSet) {
+	existingPods := util.NewStringSet()
+	for _, pod := range pods {
+		existingPods.Insert(GetPodFullName(&pod))
+	}
+	for podFullName := range mirrorPods {
+		if !existingPods.Has(podFullName) {
+			kl.mirrorManager.DeleteMirrorPod(podFullName)
+		}
+	}
+}
+
+// Helper functions.
+func getPodSource(pod *api.BoundPod) (string, error) {
+	if pod.Annotations != nil {
+		if source, ok := pod.Annotations[ConfigSourceAnnotationKey]; ok {
+			return source, nil
+		}
+	}
+	return "", errors.New(fmt.Sprintf("Cannot get source of pod %", pod.UID))
+}
+
+func isStaticPod(pod *api.BoundPod) bool {
+	source, err := getPodSource(pod)
+	if err == nil && source != ApiserverSource {
+		return true
+	}
+	return false
+}
+
+func isMirrorPod(pod *api.BoundPod) bool {
+	if value, ok := pod.Annotations[ConfigMirrorAnnotationKey]; !ok {
+		return false
+	} else {
+		return value == MirrorType
+	}
+}
+
+func parsePodFullName(podFullName string) (string, string, error) {
```

Comment:
I actually hid the function here to discourage people from using it, since most of the kubelet code is driving away from using full names. I'll move it to types.go anyway in case someone else try re-creating it again :)